File: apps/react-native/example/src/adapter/ContractMessages.ts
import { contractName } from '@banco/core-financiero';
import { messageFor } from '../contract/sources';

/**
 * Convierte un error del core en el texto que ve el usuario.
 *
 * Reusa `contractName` del paquete en vez de escribir su propio mapeo: una segunda copia se
 * desincroniza, y además el test de contrato verifica esa misma función. `contractName` y
 * `messageFor` viven las dos en `@banco/contract` —la primera reexportada desde
 * `@banco/core-financiero`, la segunda desde `../contract/sources`— así que no hay ninguna
 * lógica de mapeo propia de esta app entre el error y el string que ve el usuario.
 *
 * El `message` del binding **no se usa**: uniffi no propaga los `#[error("...")]` en español del
 * core, arma el mensaje con los campos de la variante y lo deja vacío para las que no tienen. Es
 * diagnóstico, nunca texto de usuario.
 *
 * **Los montos se interpolan crudos.** Nada de `Intl.NumberFormat` acá: los formateadores de
 * Android, iOS y el navegador no coinciden entre sí, y una diferencia rompe la comparación
 * carácter por carácter que es toda la tesis de la POC. El formateo va en el borde de
 * presentación, no acá.
 */
/**
 * Lo que ve el usuario cuando falla algo que **no** es un error de dominio. Normativo en
 * `docs/ui-spec.md`, igual en las cuatro apps. No dice «vuelve a intentarlo» a propósito: si el
 * módulo nativo no cargó, reintentar no arregla nada.
 */
export const FALLBACK = 'No se pudo completar la operación.';

export function userMessage(e: unknown): string {
  let plantilla: string;
  try {
    plantilla = messageFor(contractName(e));
  } catch {
    // **`userMessage` se llama siempre dentro de un `catch`, así que no puede lanzar.** Si lo
    // hiciera, la excepción saldría del handler del hook y llegaría al `onPress` de React: caja
    // roja en desarrollo, botón muerto en release. Y llega acá cualquier cosa que no sea un
    // `DomainError` — un `TypeError` de la capa JSI, un trap de WebAssembly, un error del
    // bundler—, porque `contractName` lanza a propósito con un tag que no reconoce (es la
    // guardia del Ruling P1, y el test de contrato **depende** de que lance: por eso el fallback
    // va acá, en el borde de UI, y no ahí).
    //
    // **El texto de diagnóstico NO va a la pantalla.** Antes acá volvía
    // `Ocurrió un error inesperado: ${String(e)}`, citando que Android hacía lo mismo — y
    // Android lo hacía mal: su test rojo mostraba `java.lang.UnsatisfiedLinkError: dlopen
    // failed: …` en la pantalla de Aritmética. La Fase 6 lo corrigió en las cuatro apps con
    // este texto, normativo en `docs/ui-spec.md`.
    //
    // El diagnóstico no se pierde: va a la consola, que es donde sirve.
    console.error('error no-dominio en el borde de UI:', e);
    return FALLBACK;
  }
  const campos = (e as { inner?: Record<string, string> }).inner ?? {};
  return plantilla.replace(/\{(\w+)\}/g, (coincidencia, clave) =>
    clave in campos ? campos[clave]! : coincidencia
  );
}

File: apps/android/core-financiero/src/main/java/dev/tohure/android_rust_test/adapter/ContractMessages.kt
package dev.tohure.android_rust_test.adapter

import dev.tohure.android_rust_test.contract.MessageSource
import uniffi.core_financiero.DomainException

/**
 * El nombre que `contracts/cases.json` le da a este error.
 *
 * `DomainError::contract_name()` es un método de Rust y **no cruza el FFI**: el enum
 * generado trae solo los nombres en inglés. Verificado en la Fase 1 con nueve variantes
 * y otra vez en la Fase 6 con la décima (`Decryption`): los diez nombres del contrato
 * aparecen 0 veces en el `.kt` generado.
 *
 * El `when` va **exhaustivo, como expresión y sin rama `else`**. Es deliberado: agregar
 * una undécima variante al core tiene que romper la compilación acá — un fallo ruidoso y
 * ubicado— en vez de caer en un `"Desconocido"` que compila, pasa en verde, y se descubre
 * el día de la demo cuando esta app muestra un error que las otras tres no.
 */
fun DomainException.contractName(): String =
    when (this) {
        is DomainException.Length -> "Longitud"
        is DomainException.CheckDigit -> "DigitoControl"
        is DomainException.UnknownBank -> "BancoDesconocido"
        is DomainException.InvalidAmount -> "MontoInvalido"
        is DomainException.AccountNotFound -> "CuentaNoEncontrada"
        is DomainException.SameAccount -> "MismaCuenta"
        is DomainException.InsufficientFunds -> "SaldoInsuficiente"
        is DomainException.Encryption -> "Cifrado"
        is DomainException.Decryption -> "Descifrado"
        is DomainException.OutOfRange -> "FueraDeRango"
    }

/**
 * Traduce un error del core al texto que ve el usuario.
 *
 * Vive en producción, no solo en el test, porque `contracts/messages.es.json` indexa los
 * mensajes por nombre del contrato: la pantalla de error necesita el mapeo igual que el
 * test de contrato. **El test de contrato reusa esta misma función** en vez de escribir la suya — así verifica
 * contra `cases.json` el mapeo que la UI usa de verdad, y no una copia que puede divergir.
 */
class ContractMessages(source: MessageSource) {
    private val messages: Map<String, String> = source.messages()

    /**
     * El texto para un `Throwable` cualquiera, que es lo que de verdad sale del adapter.
     *
     * **Esta sobrecarga no es defensiva, cubre un caso real.** `UniffiCoreFinanciero` envuelve
     * cada llamada en `runCatching`, que en Kotlin atrapa **todo `Throwable`** y no sólo
     * `DomainException`: un fallo al cargar `libcore_financiero.so`, o cualquier excepción de
     * JNA, vuelve como `Result.failure` por el mismo camino. Antes los cuatro ViewModels
     * cerraban con `?: e.toString()` y el usuario veía
     * `java.lang.UnsatisfiedLinkError: dlopen failed: …` en pantalla.
     *
     * El diagnóstico **no se pierde**: lo loguea `UniffiCoreFinanciero`, que es donde se
     * atrapa. Esta clase se queda pura —sin dependencias de Android— para que la puedan usar
     * los tests de JVM sin mockear nada.
     *
     * El texto es normativo y vive en `docs/ui-spec.md`: las cuatro apps muestran el mismo.
     */
    fun userMessage(e: Throwable): String =
        if (e is DomainException) {
            userMessage(e)
        } else {
            FALLBACK
        }

    fun userMessage(e: DomainException): String {
        val name = e.contractName()
        val template = messages[name]
            ?: error("contracts/messages.es.json no tiene el mensaje de `$name`")
        return interpolate(template, e)
    }

    /**
     * Reemplaza `{code}`, `{id}`, `{available}`, `{required}` y `{field}` por los campos
     * de la variante, **crudos**. Nada de `NumberFormat` acá: los formateadores de moneda
     * de Android, iOS y el navegador no coinciden entre sí, y una diferencia rompe la
     * comparación carácter por carácter. El formateo vive en las pantallas de montos.
     */
    private fun interpolate(template: String, e: DomainException): String =
        when (e) {
            is DomainException.Length -> template
            is DomainException.CheckDigit -> template
            is DomainException.UnknownBank -> template.replace("{code}", e.code)
            is DomainException.InvalidAmount -> template
            is DomainException.AccountNotFound -> template.replace("{id}", e.id)
            is DomainException.SameAccount -> template
            is DomainException.InsufficientFunds ->
                template
                    .replace("{available}", e.available)
                    .replace("{required}", e.required)
            is DomainException.Encryption -> template
            is DomainException.Decryption -> template
            is DomainException.OutOfRange -> template.replace("{field}", e.`field`)
        }

    companion object {
        /**
         * Lo que ve el usuario cuando falla algo que **no** es un error de dominio. Normativo
         * en `docs/ui-spec.md`, igual en las cuatro apps. No dice «vuelve a intentarlo» a
         * propósito: si la librería nativa no cargó, reintentar no arregla nada, y prometer
         * una salida que no existe es peor que no decir nada.
         */
        const val FALLBACK = "No se pudo completar la operación."
    }
}
